### PR TITLE
nixos/systemd: implement systemd.network-wait-online.ignore option

### PR DESCRIPTION
**Pull request by bendlas**
###### Description of changes

Many times when `systemctl status` says `degraded`, it will be because `systemd-networkd-wait-online.service` timed out waiting on an unused network interface.

`systemd-networkd-wait-online` does support an `--ignore` flag and this PR makes it accessible from NixOS configuration.

The target audience for this option are users for whom simply disabling the offending network interface is not an option, e.g. for laptops that are connected to an ethernet cable only sometimes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/systemd.nix [hunk 1]
```diff
@@ -606,6 +615,16 @@ in
     systemd.targets.network-online.wantedBy = [ "multi-user.target" ];
     systemd.services.systemd-importd.environment = proxy_env;
     systemd.services.systemd-pstore.wantedBy = [ "sysinit.target" ]; # see #81138
+    systemd.services.systemd-networkd-wait-online.serviceConfig.ExecStart = mkIf
+      (0 != length config.systemd.network-wait-online.ignore) [
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd.nix:619]: (nit) `config.systemd.network-wait-online.ignore != []` may be a little bit more lightweight to evaluate.

File: nixos/modules/system/boot/systemd.nix [hunk 2]
```diff
@@ -606,6 +615,16 @@ in
     systemd.targets.network-online.wantedBy = [ "multi-user.target" ];
     systemd.services.systemd-importd.environment = proxy_env;
     systemd.services.systemd-pstore.wantedBy = [ "sysinit.target" ]; # see #81138
+    systemd.services.systemd-networkd-wait-online.serviceConfig.ExecStart = mkIf
```
Comment (Ma27) [inline nixos/modules/system/boot/systemd.nix:618]: I guess it's better to use `systemd.network.wait-online.extraArgs` to make sure that other arguments configured aren't silently discarded.

**Review by Ma27 — COMMENTED** [review]
Just in case you don't know already, `systemd.network.wait-online.anyInterface = true;` may also fix your issue.

But anyways, considering that we already expose `--any` over the module system, I'd be fine with doing the same for `--ignore`.

**Comment by ElvishJerricco** [comment]
I find this a little odd. We don't enable the `wait-online` service unless explicitly configured by the user; so this isn't solving a general problem. When the user enables `wait-online`, it's because they want its behavior. So ignoring that behavior some of the time seems odd to me.

**Comment by Ma27** [comment]
`systemd.network.wait-online.enable` is `true` by default.

**Comment by bendlas** [comment]
> Just in case you don't know already, `systemd.network.wait-online.anyInterface = true;` may also fix your issue.

Or maybe [`systemd.network.wait-online.ignoredInterfaces`](https://github.com/NixOS/nixpkgs/blob/3ae20aa58a6c0d1ca95c9b11f59a2d12eebc511f/nixos/modules/system/boot/networkd.nix#L3030) might? 🙃

Sorry for wasting your time on this review. This has been sitting in my outbox queue so long that apparently we already grew that option.

**Comment by ElvishJerricco** [comment]
> `systemd.network.wait-online.enable` is `true` by default

Huh. Don't know how I had that wrong in my head. Could have sworn that wasn't the case. Apologies. That completely invalidates my comment :P

**Pull request closed without merging**